File: lib/blitz3d/translator/node/restore.rb
module Blitz3D
  module AST
    class RestoreNode < StmtNode
      attr_accessor :ident, :sem_label

      def initialize(json)
        super
        @ident = json['ident']
        @sem_label = Label.new(json['sem_label'])
      end

      def to_c
        "_bbRestore( &_DATA[#{sem_label.data_sz}] )"
      end
    end
  end
end

File: src/tools/compiler/tree/stmt/restore.cpp
#include "restore.h"

//////////////////
// Restore data //
//////////////////
void RestoreNode::semant( Environ *e ){
	if( e->level>0 ) e=e->globals;

	if( ident.size()==0 ) sem_label=0;
	else{
		sem_label=e->findLabel( ident );
		if( !sem_label ) sem_label=e->insertLabel( ident,-1,pos,-1 );
	}
}

void RestoreNode::translate( Codegen *g ){
	TNode *t=global( "__DATA" );
	if( sem_label ) t=add( t,iconst( sem_label->data_sz*8 ) );
	g->code( call( "__bbRestore",t ) );
}

#ifdef USE_LLVM
void RestoreNode::translate2( Codegen_LLVM *g ){
	g->restoreData( sem_label?sem_label->data_sz:0 );
}
#endif
